File: frontend/src/components/benchmark/content/common_neighbors/implementations/common_neighbors.java
import java.util.*;

public class CommonNeighbors {
    private Map<Integer, Set<Integer>> adjacencyList;

    public CommonNeighbors() {
        this.adjacencyList = new HashMap<>();
    }

    public void addEdge(int u, int v) {
        adjacencyList.putIfAbsent(u, new HashSet<>());
        adjacencyList.putIfAbsent(v, new HashSet<>());
        adjacencyList.get(u).add(v);
        adjacencyList.get(v).add(u);
    }

    public void buildFromEdges(List<int[]> edges) {
        for (int[] edge : edges) {
            addEdge(edge[0], edge[1]);
        }
    }

    public Set<Integer> commonNeighbors(int u, int v) {
        if (!adjacencyList.containsKey(u) || !adjacencyList.containsKey(v)) {
            return new HashSet<>();
        }

        Set<Integer> result = new HashSet<>();
        for (int node : adjacencyList.get(u)) {
            if (adjacencyList.get(v).contains(node)) {
                result.add(node);
            }
        }
        return result;
    }

    public int commonNeighborsCount(int u, int v) {
        return commonNeighbors(u, v).size();
    }

    public boolean predictLink(int u, int v, int threshold) {
        return commonNeighborsCount(u, v) >= threshold;
    }

    public List<Map.Entry<int[], Integer>> scoreAllPairs() {
        List<Map.Entry<int[], Integer>> scores = new ArrayList<>();
        List<Integer> nodes = new ArrayList<>(adjacencyList.keySet());

        for (int i = 0; i < nodes.size(); i++) {
            for (int j = i + 1; j < nodes.size(); j++) {
                int u = nodes.get(i);
                int v = nodes.get(j);
                int count = commonNeighborsCount(u, v);
                if (count > 0) {
                    scores.add(new AbstractMap.SimpleEntry<>(new int[]{u, v}, count));
                }
            }
        }

        scores.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        return scores;
    }

    public List<Map.Entry<int[], Integer>> topKPredictions(int k) {
        List<Map.Entry<int[], Integer>> scores = scoreAllPairs();
        return scores.subList(0, Math.min(k, scores.size()));
    }
}

File: frontend/src/components/benchmark/content/common_neighbors/implementations/common_neighbors.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class CommonNeighbors
{
    private Dictionary<int, HashSet<int>> adjacencyList;

    public CommonNeighbors()
    {
        adjacencyList = new Dictionary<int, HashSet<int>>();
    }

    public void AddEdge(int u, int v)
    {
        if (!adjacencyList.ContainsKey(u))
            adjacencyList[u] = new HashSet<int>();
        if (!adjacencyList.ContainsKey(v))
            adjacencyList[v] = new HashSet<int>();

        adjacencyList[u].Add(v);
        adjacencyList[v].Add(u);
    }

    public void BuildFromEdges(List<(int, int)> edges)
    {
        foreach (var (u, v) in edges)
        {
            AddEdge(u, v);
        }
    }

    public HashSet<int> CommonNeighbors(int u, int v)
    {
        if (!adjacencyList.ContainsKey(u) || !adjacencyList.ContainsKey(v))
            return new HashSet<int>();

        var result = new HashSet<int>();
        foreach (var node in adjacencyList[u])
        {
            if (adjacencyList[v].Contains(node))
                result.Add(node);
        }
        return result;
    }

    public int CommonNeighborsCount(int u, int v)
    {
        return CommonNeighbors(u, v).Count;
    }

    public bool PredictLink(int u, int v, int threshold = 1)
    {
        return CommonNeighborsCount(u, v) >= threshold;
    }

    public List<((int, int), int)> ScoreAllPairs()
    {
        var scores = new List<((int, int), int)>();
        var nodes = adjacencyList.Keys.ToList();

        for (int i = 0; i < nodes.Count; i++)
        {
            for (int j = i + 1; j < nodes.Count; j++)
            {
                int u = nodes[i];
                int v = nodes[j];
                int count = CommonNeighborsCount(u, v);
                if (count > 0)
                {
                    scores.Add(((u, v), count));
                }
            }
        }

        return scores.OrderByDescending(x => x.Item2).ToList();
    }

    public List<((int, int), int)> TopKPredictions(int k = 10)
    {
        return ScoreAllPairs().Take(k).ToList();
    }
}